internal/osutils/lockfile/lock_test.go

```diff
@@ -0,0 +1,193 @@
+package lockfile
+
+import (
+	"context"
+	"errors"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+
+	"io/ioutil"
+
+	"github.com/ActiveState/cli/internal/environment"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func buildTestExecutable(t *testing.T, dir string) string {
+	root, err := environment.GetRootPath()
+	require.NoError(t, err)
+	lockerExe := filepath.Join(dir, "locker")
+	if runtime.GOOS == "windows" {
+		lockerExe += ".exe"
+	}
+
+	cmd := exec.Command(
+		"go", "build", "-o", lockerExe,
+		filepath.Join(root, "internal", "osutils", "testdata", "locker"),
+	)
+	err = cmd.Run()
+	require.NoError(t, err)
+
+	return lockerExe
+}
+
+func Test_acquirePidLockProcesses(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	// build the test process that acquires a lock
+	lockerExe := buildTestExecutable(t, tmpDir)
+
+	t.Run("locked in other process", func(tt *testing.T) {
+		lockFile := filepath.Join(tmpDir, "locked-")
+
+		lockCmd := exec.Command(lockerExe, lockFile)
+		lockCmd = prepLockCmd(lockCmd)
+
+		stdout, err := lockCmd.StdoutPipe()
+		require.NoError(tt, err)
+		err = lockCmd.Start()
+		require.NoError(tt, err)
+
+		// wait for command to block
+		buf := make([]byte, 6)
+		n, err := stdout.Read(buf)
+		require.NoError(tt, err)
+		require.Equal(tt, 6, n)
+		assert.Equal(tt, "LOCKED", string(buf))
+
+		pl, err := NewLock(lockFile)
+		require.NoError(tt, err)
+
+		// trying to acquire the lock in this process should fail
+		err = pl.TryLock()
+		require.Error(tt, err)
+		alreadyErr := &AlreadyLockedError{}
+		assert.True(tt, errors.As(err, &alreadyErr))
+
+		err = pl.Close()
+		require.NoError(tt, err)
+
+		// stopping the other process
+		interruptProcess(tt, lockCmd.Process)
+
+		// waiting for the process to finish without error
+		err = lockCmd.Wait()
+		require.NoError(tt, err)
+		assert.True(tt, lockCmd.ProcessState.Exited())
+		assert.Equal(tt, 0, lockCmd.ProcessState.ExitCode())
+	})
+
+	t.Run("stress-test", func(tt *testing.T) {
+		// stress tests runs numProcesses in parallel, and only one should get the lock
+		numProcesses := 10
+
+		lockFile := filepath.Join(tmpDir, "stress-test")
+
+		done := make(chan string, numProcesses+1)
+		defer close(done)
+		var wg sync.WaitGroup
+		defer wg.Wait()
```

Comment:
not nit, but nit for now: This is just so janky that I need to note it. Maybe we can talk about this sort of thing in architecture talks at some point.